src/AIS.cpp

```diff
@@ -129,3 +131,113 @@ ais_base::VoyageInformation AIS::getVoyageInformation(ais::message_05 const& mes
+        return {Eigen::Quaterniond(
+                    Eigen::AngleAxisd(yaw.value().getRad(), Eigen::Vector3d::UnitZ())),
+            ais_base::PositionCorrectionStatus::POSITION_CENTERED_USING_HEADING};
+    }
+    else if (course_over_ground.has_value() &&
+             !std::isnan(course_over_ground.value().getDeg()) &&
+             speed_over_ground >= MIN_SPEED_THRESHOLD) {
+        return {Eigen::Quaterniond(Eigen::AngleAxisd(course_over_ground.value().getRad(),
+                    Eigen::Vector3d::UnitZ())),
+            ais_base::PositionCorrectionStatus::POSITION_CENTERED_USING_COURSE};
+    }
+    else {
+        return {Eigen::Quaterniond::Identity(),
+            ais_base::PositionCorrectionStatus::POSITION_RAW};
+    }
+}
+
+base::Vector3d sensorToVesselInWorldPose(base::Vector3d sensor2vessel_pos,
+    Eigen::Quaterniond vessel2world_ori)
+{
+    base::Vector3d sensor2vessel_in_world_pos;
+    sensor2vessel_in_world_pos = vessel2world_ori * sensor2vessel_pos;
+
+    return sensor2vessel_in_world_pos;
+}
+
+base::samples::RigidBodyState convertGPSToUTM(
+    const std::optional<ais_base::Position>& position,
+    gps_base::UTMConverter utm_converter)
+{
+    if (!position.has_value()) {
+        std::string error_msg = "Position data is unavailable";
+        LOG_ERROR_S << error_msg;
+        throw std::invalid_argument(error_msg);
+    }
+
+    auto position_value = position.value();
+    gps_base::Solution sensor2world_solution;
+    sensor2world_solution.latitude = position_value.latitude.getDeg();
+    sensor2world_solution.longitude = position_value.longitude.getDeg();
+
+    base::samples::RigidBodyState sensor2world_UTM;
+    sensor2world_UTM.position =
+        utm_converter.convertToUTM(sensor2world_solution).position;
+    sensor2world_UTM.time = base::Time::now();
+
+    return sensor2world_UTM;
+}
+
+std::pair<base::Angle, base::Angle> convertUTMToGPSInWorldFrame(
+    base::samples::RigidBodyState sensor2world_UTM,
+    base::Vector3d sensor2vessel_in_world_pos,
+    gps_base::UTMConverter utm_converter)
+{
+    base::samples::RigidBodyState vessel2world_UTM;
+    vessel2world_UTM.position = sensor2world_UTM.position + sensor2vessel_in_world_pos;
+
+    gps_base::Solution vessel2world_GPS;
+    vessel2world_GPS = utm_converter.convertUTMToGPS(vessel2world_UTM);
+
+    return {base::Angle::fromDeg(vessel2world_GPS.latitude),
+        base::Angle::fromDeg(vessel2world_GPS.longitude)};
+}
+
+ais_base::Position AIS::applyPositionCorrection(ais_base::Position const& position,
+    base::Vector3d const& vessel_reference_position,
+    gps_base::UTMConverter utm_converter)
+{
+    if (std::isnan(position.yaw.getRad()) &&
+        std::isnan(position.course_over_ground.getRad())) {
+        std::string error_msg = "Position can't be corrected because both 'yaw' "
+                                "and 'course_over_ground' values are missing.";
+        LOG_ERROR_S << error_msg << std::endl;
```

Comment:
Is there a point in storing the error into a string before logging it out?
```suggestion
        LOG_ERROR_S <<  "Position can't be corrected because both 'yaw' ...";
```

Reply:
There was before, when I was also throwing an error. Will fix it.